audit(vscode-extension): pre-announcement quality state

## Summary

Pre-announcement quality audit of `vscode-extension/` for v0.13.0. Overall the extension is in solid shape: zero `any` types in production source, zero `console.*` calls, zero TODO/FIXME markers, 397 test cases across 15 jest files, complete `contributes` section, valid icon, full `repository`/`bugs`/`homepage` metadata, gitignored `out/` and `*.vsix`, and a comprehensive README.

ESLint v10.2.0 (post #3179) reports **0 errors / 7 warnings**, all minor.

This issue documents the gaps to fix before the public alpha announcement, sorted by severity.

## Per-area findings

### Manifest (`package.json`) — mostly good

Verified:
- `version: 0.12.2` matches workspace
- `engines.vscode: ^1.88.0` (April 2024) — reasonable, not stale
- `categories`: `Programming Languages`, `Linters`, `Formatters`, `Debuggers`, `Testing` — appropriate
- `keywords`: 10 marketplace terms including `vscodium` and `open-vsx` — good coverage
- `icon`: `icon.png` exists (256x256 RGBA PNG, 8.4KB)
- `repository`, `bugs`, `homepage` — all set to `EffortlessMetrics/perl-lsp`
- `license: MIT`, `publisher: EffortlessMetrics`, `preview: true`
- `extensionKind: ["workspace"]`, `capabilities` declares virtual + untrusted workspace support
- `contributes`: languages, grammars, breakpoints, debuggers, configuration (4 sections, 19 settings), commands (24), menus (commandPalette/editor/title/editor/context), keybindings (5), snippets (3), taskDefinitions, walkthroughs (7 steps)
- `scripts`: `lint`, `test`, `vscode:prepublish`, `compile`, `package`, `publish`, `publish:openvsx` all defined

### Lint (`pnpm run lint`) — 0 errors, 7 warnings

```
src/debugAdapter.ts
  359:9  warning  'session' is defined but never used         no-unused-vars
  360:9  warning  'executable' is defined but never used      no-unused-vars
  461:9  warning  'token' is defined but never used           no-unused-vars
  497:9  warning  'folder' is defined but never used          no-unused-vars
  498:9  warning  'token' is defined but never used           no-unused-vars

src/extension.ts
  5:1    warning  Imports "LanguageClientOptions" and "ServerOptions" are only used as type
                                                              consistent-type-imports

src/onboarding.ts
  63:3   warning  Unused eslint-disable directive (no problems were reported from
                  '@typescript-eslint/naming-convention')
```

Breakdown: 5 unused-args (rename to `_session`, `_executable`, `_token`, `_folder`, `_token`), 1 type-only import, 1 stale eslint-disable directive. 2 are auto-fixable with `eslint --fix`.

### README — comprehensive, has version drift

Strong:
- Marketplace + Open VSX badges (download counts)
- Full feature sections (Navigation, Refactoring, Diagnostics, Advanced, Debugging, Test Explorer)
- Installation instructions for VS Code, VSCodium, Homebrew, curl, cargo
- Configuration table (15 settings documented)
- Keyboard shortcuts table
- Per-editor compatibility matrix (VS Code, Cursor, Neovim, Emacs, Helix, Sublime, Codespaces, Gitpod)
- Troubleshooting section
- Air-gapped/internal deployment pointer

**Issue (HIGH)**: README line 10 says `> **0.12.1 Public Alpha**` but `package.json` is `0.12.2`. Version drift visible to marketplace shoppers.

**Issue (LOW)**: Walkthrough storyboard SVGs (welcome, verify-perl, open-project, try-completion, try-goto-definition, configure-settings, debug-first-script) are static SVGs, not animated GIFs. README acknowledges this ("not the final recorded GIFs"). Acceptable for alpha but the v0.13.0 announcement may want at least a GIF or screenshot showing actual extension behavior in the marketplace listing.

### Tests — substantial real coverage

- 15 test files, 4,074 lines, **397 test cases**
- Covers: arrow completion, command registration, configuration validation, debug adapter, downloader (logic + integration), eslint config, file creation, formatting errors, health widget, language client args, onboarding, POD preview, streaming completion, walkthrough, what's new
- Real assertions, not stubs (downloader.test.ts alone has 759 lines and 63 cases including spy-based logic verification)
- `jest.config.js` uses `ts-jest` preset with mocked `vscode` module

No quality gap here. Tests not run in this audit (orchestrator did not request CI invocation), but the structure is sound.

### Dead code / `any` types — clean in production

- `grep -nE ': any|<any>|as any|any\[\]' src/*.ts` (production files only): **0 matches**
- `grep -nE 'TODO|FIXME|XXX|HACK' src/*.ts`: **0 matches**
- `grep -nE 'console\.' src/*.ts`: **0 matches**
- All `any` usage is confined to `src/test/**` (mocks, jest spies on private members, package.json walking) — reasonable test patterns

#3154 cleanup verified: no new `any` regressions in production source.

### Other findings

**`@types/node: ^25.5.2`** (HIGH-MEDIUM) — Node 25 is bleeding-edge. VSCode 1.88 ships Electron with Node 18.x; this means TypeScript will accept APIs that don't exist at runtime in the actual VSCode host. Recommend pinning to `@types/node: ^18.x` or `^20.x` to match the VSCode runtime. This is a latent footgun, not a current bug.

**`activationEvents` is internally inconsistent** (LOW) — VSCode >= 1.75 auto-activates on declared `commands` so explicit `onCommand:*` entries are redundant. The list declares 14 explicit `onCommand:*` entries but **omits** at least 8 commands that ARE registered: `showOutput`, `showStatusMenu`, `organizeImports`, `runTests`, `extractVariable`, `extractMethod`, `showRefactoringOptions`, `showWhatsNew`, `reportIssue`. Either drop the explicit list (rely on auto-activation, recommended) or make it complete. The current half-state is misleading.

**`.vscodeignore` is correct** — excludes `out/test/**`, `coverage/**`, `src/**`, `test/**`, `*.vsix`, `*.tgz`, `node_modules/**`, scripts, build files, dev docs. Allowlists `README.md`, `CHANGELOG.md`, `LICENSE`. Good.

**CHANGELOG drift** (HIGH) — `CHANGELOG.md` last entry is `[0.12.1] - 2026-03-30`. There is **no entry for 0.12.2**. The marketplace listing displays the changelog; users updating from 0.12.1 to 0.12.2 will see no notes for the change.

**Stale build artifact** (LOW, gitignored) — `out/` directory on disk is from Apr 2 and missing `streamingCompletion.js` even though `streamingCompletion.ts` is imported in `extension.ts`. `out/` is gitignored so this only affects local dev — `pnpm run compile` regenerates it. Note for builders: don't ship a `.vsix` without first running `pnpm run compile`. Also `perl-lsp-rs-0.12.0.vsix` (4.4MB, version-stale) is in the worktree but gitignored — left over from a previous package run, harmless but worth deleting from local dirs.

**`bin/`** — gitignored, present locally; expected runtime download target.

## Recommendations (sorted by severity)

### HIGH — fix before announcement

1. **Add 0.12.2 entry to `vscode-extension/CHANGELOG.md`** documenting what shipped (eslint v10 bump #3179, plus any other 0.12.2 changes since 0.12.1).
2. **Update README line 10** from `**0.12.1 Public Alpha**` to `**0.12.2 Public Alpha**` (or, better, drop the version number from the prose and point at the badge / CHANGELOG so this never drifts again).
3. **Pin `@types/node`** to `^18.x` or `^20.x` to match VSCode 1.88's actual Node runtime. Current `^25.5.2` allows TS code to call APIs that won't exist at runtime.

### MEDIUM — fix this cycle

4. **Resolve the 7 lint warnings**:
   - In `src/debugAdapter.ts`: rename `session` (359), `executable` (360), `token` (461), `folder` (497), `token` (498) to `_session`, `_executable`, `_token`, `_folder`, `_token`.
   - In `src/extension.ts:5`: convert `LanguageClientOptions` and `ServerOptions` to `import type { ... }` (auto-fixable).
   - In `src/onboarding.ts:63`: remove the dead `eslint-disable` directive (auto-fixable).
5. **Decide on `activationEvents`**: either drop the explicit `onCommand:*` list (recommended for VSCode >= 1.75 — auto-activation handles it) or add the missing 8+ commands. Half-list is technical debt.

### LOW — nice to have for v0.13.0

6. **Add at least one real screenshot or animated GIF** to the README showing go-to-definition, hover, or completion in action. The current walkthrough SVGs are storyboards, not product shots, and the marketplace listing is stronger with a real screenshot.
7. **Clean up local dev artifacts**: delete the stale `perl-lsp-rs-0.12.0.vsix` from working trees (it's gitignored, harmless, just clutter).

## Reproduction commands

```bash
cd vscode-extension
pnpm install --frozen-lockfile
pnpm run lint  # 0 errors, 7 warnings
grep -nE ': any|<any>|as any|any\[\]' src/*.ts  # 0 matches in production
grep -n "0\.12" README.md  # finds the 0.12.1 drift
grep "0\.12\.2" CHANGELOG.md  # 0 matches
```

## Boundaries

- Audit was read-only
- `pnpm install --frozen-lockfile` did not modify the lockfile
- `pnpm run lint` is idempotent
- `git status` clean after audit
- No source files modified

--- a/vscode-extension/CHANGELOG.md
+++ b/vscode-extension/CHANGELOG.md
@@ -2,6 +2,25 @@
 
 All notable changes to the Perl Language Server extension will be documented in this file.
 
+## [0.12.3] - 2026-04-09
+
+The 0.12.3 release line. Aligned with the workspace `v0.12.3` cut and consolidates
+the user-facing UX hardening work from April 2026.
+
+### Fixed
+- **Actionable error messages**: server startup errors now surface specific,
+  actionable messages instead of a generic "corrupted" fallback
+  (PR #3280, PR #3291).
+- **Perl interpreter detection**: distinct error shown when the Perl interpreter
+  is not found in PATH, with a clear install prompt (PR #3275).
+- **Binary download errors**: download failures include the underlying OS error
+  and a retry hint instead of a silent failure (PR #3274).
+- **Settings schema polish**: extension settings entries cleaned up for the
+  VS Code Settings UI — descriptions, defaults, and enum labels improved
+  (PR #3269).
+- **Enterprise/offline deployment**: documentation added for air-gapped binary
+  distribution and internal mirror configuration (PR #3276).
+
 ## [0.12.2] - 2026-04-04
 
 The 0.12.2 release line. Aligned with the workspace `v0.12.2` cut on 2026-04-04

--- a/vscode-extension/CHANGELOG.md
+++ b/vscode-extension/CHANGELOG.md
@@ -10,16 +10,16 @@ the user-facing UX hardening work from April 2026.
 ### Fixed
 - **Actionable error messages**: server startup errors now surface specific,
   actionable messages instead of a generic "corrupted" fallback
-  (PR #3280, PR #3291).
+  (PR #3308, PR #3291).
 - **Perl interpreter detection**: distinct error shown when the Perl interpreter
-  is not found in PATH, with a clear install prompt (PR #3275).
+  is not found in PATH, with a clear install prompt (PR #3312).
 - **Binary download errors**: download failures include the underlying OS error
-  and a retry hint instead of a silent failure (PR #3274).
+  and a retry hint instead of a silent failure (PR #3306).
 - **Settings schema polish**: extension settings entries cleaned up for the
   VS Code Settings UI — descriptions, defaults, and enum labels improved
   (PR #3269).
 - **Enterprise/offline deployment**: documentation added for air-gapped binary
-  distribution and internal mirror configuration (PR #3276).
+  distribution and internal mirror configuration (PR #3310).
 
 ## [0.12.2] - 2026-04-04
 

--- a/vscode-extension/README.md
+++ b/vscode-extension/README.md
@@ -7,7 +7,7 @@
 
 A fast, native Perl 5 language server extension. Written in Rust for speed and reliability. No runtime dependencies -- just install and code.
 
-> **0.12.2 Public Alpha** -- This extension is under active development. Please [report issues](https://github.com/EffortlessMetrics/perl-lsp/issues/new/choose) if you encounter problems.
+> **0.12.3 Public Alpha** -- This extension is under active development. Please [report issues](https://github.com/EffortlessMetrics/perl-lsp/issues/new/choose) if you encounter problems.
 
 ## Features
 

--- a/vscode-extension/package.json
+++ b/vscode-extension/package.json
@@ -54,21 +54,6 @@
   "main": "./out/extension.js",
   "activationEvents": [
     "onLanguage:perl",
-    "onCommand:perl-lsp.restart",
-    "onCommand:perl-lsp.showVersion",
-    "onCommand:perl-lsp.reinstall",
-    "onCommand:perl-lsp.checkSyntax",
-    "onCommand:perl-lsp.runCurrentTest",
-    "onCommand:perl-lsp.runAllTests",
-    "onCommand:perl-lsp.formatDocument",
-    "onCommand:perl-lsp.showIncPaths",
-    "onCommand:perl-lsp.openModule",
-    "onCommand:perl-lsp.showParserAst",
-    "onCommand:perl-lsp.checkForUpdate",
-    "onCommand:perl-lsp.createDebugConfig",
-    "onCommand:perl-lsp.runHealthCheck",
-    "onCommand:perl-lsp.openConfigurationGuide",
-    "onCommand:perl-lsp.previewPod",
     "onWalkthrough:perl-lsp.gettingStarted"
   ],
   "contributes": {
@@ -866,7 +851,7 @@
   "devDependencies": {
     "@types/adm-zip": "^0.5.8",
     "@types/jest": "^30.0.0",
-    "@types/node": "^25.5.2",
+    "@types/node": "^20.x",
     "@types/tar": "^7.0.87",
     "@types/vscode": "^1.88.0",
     "@typescript-eslint/eslint-plugin": "^8.58.0",

--- a/vscode-extension/src/debugAdapter.ts
+++ b/vscode-extension/src/debugAdapter.ts
@@ -356,8 +356,8 @@ export class PerlDebugAdapterDescriptorFactory implements vscode.DebugAdapterDes
     constructor(private readonly context: vscode.ExtensionContext) {}
 
     createDebugAdapterDescriptor(
-        session: vscode.DebugSession,
-        executable: vscode.DebugAdapterExecutable | undefined
+        _session: vscode.DebugSession,
+        _executable: vscode.DebugAdapterExecutable | undefined
     ): vscode.ProviderResult<vscode.DebugAdapterDescriptor> {
         // Try to find perl-dap in PATH or use bundled version
         const dapPath = this.findDebugAdapter();
@@ -463,9 +463,9 @@ export class PerlDebugAdapterDescriptorFactory implements vscode.DebugAdapterDes
 
 export class PerlDebugConfigurationProvider implements vscode.DebugConfigurationProvider {
     resolveDebugConfiguration(
-        folder: vscode.WorkspaceFolder | undefined,
+        _folder: vscode.WorkspaceFolder | undefined,
         config: vscode.DebugConfiguration,
-        token?: vscode.CancellationToken
+        _token?: vscode.CancellationToken
     ): vscode.ProviderResult<vscode.DebugConfiguration> {
         // If launch.json is missing or empty
         if (!config.type && !config.request && !config.name) {
@@ -501,8 +501,8 @@ export class PerlDebugConfigurationProvider implements vscode.DebugConfiguration
     }
 
     provideDebugConfigurations(
-        folder: vscode.WorkspaceFolder | undefined,
-        token?: vscode.CancellationToken
+        _folder: vscode.WorkspaceFolder | undefined,
+        _token?: vscode.CancellationToken
     ): vscode.ProviderResult<vscode.DebugConfiguration[]> {
         return [
             {

--- a/vscode-extension/src/extension.ts
+++ b/vscode-extension/src/extension.ts
@@ -2,13 +2,8 @@ import * as vscode from 'vscode';
 import * as path from 'path';
 import * as fs from 'fs';
 import { execFile } from 'child_process';
-import {
-    LanguageClient,
-    LanguageClientOptions,
-    ServerOptions,
-    TransportKind,
-    Trace
-} from 'vscode-languageclient/node';
+import { LanguageClient, TransportKind, Trace } from 'vscode-languageclient/node';
+import type { LanguageClientOptions, ServerOptions } from 'vscode-languageclient/node';
 import { PerlTestAdapter } from './testAdapter';
 import { activateDebugger, rewriteDebugTestLensCommand } from './debugAdapter';
 import { BinaryDownloader } from './downloader';

--- a/vscode-extension/src/onboarding.ts
+++ b/vscode-extension/src/onboarding.ts
@@ -109,7 +109,6 @@ export class OnboardingManager {
    * Replaceable exec function.  In production this wraps `child_process.execFile`;
    * in tests it is replaced with a jest mock via `mgr._execCheck = jest.fn(...)`.
    */
-  // eslint-disable-next-line @typescript-eslint/naming-convention
   _execCheck: ExecCheckFn;
 
   constructor(

--- a/vscode-extension/src/test/onboarding.test.ts
+++ b/vscode-extension/src/test/onboarding.test.ts
@@ -307,11 +307,14 @@ describe('package.json health check command', () => {
     expect(cmd.title.toLowerCase()).toContain('health');
   });
 
-  test('runHealthCheck has an activation event so it works without a Perl file open', () => {
+  test('runHealthCheck is declared as a command so VSCode auto-activates the extension', () => {
     // runHealthCheck is palette-global (no when clause restricting to editorLangId == perl).
-    // Without its own activation event, VSCode will not load the extension when the user
-    // triggers the command from the command palette if no Perl file is active.
-    expect(pkg.activationEvents).toContain('onCommand:perl-lsp.runHealthCheck');
+    // VSCode >= 1.75 automatically activates an extension when any of its declared commands are
+    // triggered — explicit onCommand:* activationEvents entries are redundant and have been
+    // removed. The guarantee is that the command exists in contributes.commands.
+    const commands = pkg.contributes.commands as Array<{ command: string; title: string }>;
+    const cmd = commands.find((c) => c.command === 'perl-lsp.runHealthCheck');
+    expect(cmd).toBeDefined();
   });
 
   test('runHealthCheck is listed in commandPalette without a language restriction', () => {

